### PR TITLE
Upgrade micronaut 2 to 3

### PR DESCRIPTION
**Pull request by kevintanhongann**
Changes:
1. upgraded micronaut 2.4.4 to 3.6.0
2. upgraded micronaut data to 3.7.2 to support IgnoreCase queries
3. changed javax.inject packages to jakarta.inject
4. rxjava3 gradle dependency added for less code changes.
5. Auth classes and test classes fixed to the newest version based on the breaking changes migrating from micronaut 2 to 3.
6. defaulted the server port for dev, prod and test to 8080

Closes issue #259

### CODE REVIEW

**Comment by atomfrede** [comment]
👍 impressive work. Awesome.

**Comment by kevintanhongann** [comment]
@atomfrede @mraible are there any things on the npm side that I should be aware of when I do PRs?

File: package.json
Comment (mraible) [inline package.json:52]: I don't think you want to remove this. I'd add it back in and see if the CI tests pass. If they do, try upgrading it to 7.9.2 (the latest release).

**Review by mraible — CHANGES_REQUESTED** [review]

File: package.json
Comment (kevintanhongann) [reply package.json:52]: Okay got it. I'll try again.

File: generators/server/templates/gradle.properties.ejs
Comment (atomfrede) [inline generators/server/templates/gradle.properties.ejs:68]: Should this be updated too?

File: generators/server/templates/gradle.properties.ejs
Comment (atomfrede) [inline generators/server/templates/gradle.properties.ejs:71]: lets update to 3.4.0, which supports e.g. arm64 and has other improvements

File: generators/server/templates/gradle.properties.ejs
Comment (atomfrede) [inline generators/server/templates/gradle.properties.ejs:81]: Maybe also update to 7.x?

**Review by atomfrede — COMMENTED** [review]
Minor comments, regarding other dependencies and version updates

**Comment by atomfrede** [comment]
In the [ci configuration you might update to jhipster 7.9.2 ](https://github.com/jhipster/generator-jhipster-micronaut/blob/be65ac545f262977bb77b6c5dcce667dc02ea3f5/.github/workflows/samples.yml#L62)

**Comment by kevintanhongann** [comment]
Okay, I'll look into it.

File: generators/constants.cjs
Comment (kevintanhongann) [reply generators/constants.cjs:51]: Lemme know what's the latest version of jackson and logback as well.

Comment (atomfrede) [reply generators/constants.cjs:51]: logback should be 1.2.11 and jackson 2.13.3

File: generators/server/templates/gradle.properties.ejs
Comment (kevintanhongann) [reply generators/server/templates/gradle.properties.ejs:71]: done.

File: generators/constants.cjs
Comment (kevintanhongann) [reply generators/constants.cjs]: done

File: generators/server/templates/gradle.properties.ejs
Comment (kevintanhongann) [reply generators/server/templates/gradle.properties.ejs:68]: I have no idea. But I can try to see if it succeeds or fails

File: generators/server/templates/gradle.properties.ejs
Comment (atomfrede) [inline generators/server/templates/gradle.properties.ejs:69]: Sure this version exists?

**Comment by atomfrede** [comment]
I was wrong we have initial microservice support. So fixing the server port to 8080 will not work.

**Comment by atomfrede** [comment]
Failing tests look like wrong jdk version. We support jdk11 by default, but of course we can only support jdk17 in this blueprint if needed.

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project jhipster: Fatal error compiling: java.lang.NoSuchMethodError: 'void io.micronaut.annotation.processing.visitor.JavaClassElement.<init>(javax.lang.model.element.TypeElement, io.micronaut.core.annotation.AnnotationMetadata, io.micronaut.annotation.processing.visitor.JavaVisitorContext, java.util.Map, int)' -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error: Process completed with exit code 1.
```

**Comment by kevintanhongann** [comment]
> Failing tests look like wrong jdk version. We support jdk11 by default, but of course we can only support jdk17 in this blueprint if needed.
> 
> ```
> Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project jhipster: Fatal error compiling: java.lang.NoSuchMethodError: 'void io.micronaut.annotation.processing.visitor.JavaClassElement.<init>(javax.lang.model.element.TypeElement, io.micronaut.core.annotation.AnnotationMetadata, io.micronaut.annotation.processing.visitor.JavaVisitorContext, java.util.Map, int)' -> [Help 1]
> Error:  
> Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
> Error:  Re-run Maven using the -X switch to enable full debug logging.
> Error:  
> Error:  For more information about the errors and possible solutions, please read the following articles:
> Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
> Error: Process completed with exit code 1.
> ```

I suppose this is set within github actions samples.yml? The temurin (jdk17) is the version name. Looks like we need to change that.

**Comment by atomfrede** [comment]
Its set from main generator. But you can set it fixed to 17 here in the gh action: https://github.com/jhipster/generator-jhipster-micronaut/blob/main/.github/workflows/samples.yml#L76

File: .github/workflows/samples.yml
Comment (atomfrede) [inline .github/workflows/samples.yml:76]: It was 11 before

**Comment by kevintanhongann** [comment]
Looks like I haven't fixed the maven portion of the generator.

File: .github/workflows/samples.yml
Comment (atomfrede) [inline .github/workflows/samples.yml:76]:
```suggestion
          java-version: '17'
```

**Comment by atomfrede** [comment]
I think we need to do some changes to generated pom.xml (plugin update, enforcer rule update...)

**Comment by kevintanhongann** [comment]
I'm handling that right now. I've generated a sample maven project to spot
the differences for upgrading.

**Comment by atomfrede** [comment]
I will try to make some changes on your branch if thats okay.

**Comment by kevintanhongann** [comment]
> I will try to make some changes on your branch if thats okay.

not to say that is not, but let me finish what I started so that I can get to learn something about the maven side of the generator as well. Besides, there are also some other issues to fix. If there are any questions, I will definitely ask for your help if I'm stuck.

**Comment by atomfrede** [comment]
Sorry, was too fast. Just fixed the maven enforcer rule to allow building with jdk17

**Comment by kevintanhongann** [comment]
> Sorry, was too fast. Just fixed the maven enforcer rule to allow building with jdk17

no worries, thanks.

**Comment by kevintanhongann** [comment]
#174 i think one of the reasons the CI is affected because of this issue. I don't know what is the status of the issue, but please do let me know ASAP on what we can do about it. @atomfrede @mraible

**Comment by atomfrede** [comment]
CI for your branch is not affected by the other issues or PR. Or did I get you wrong?

**Comment by kevintanhongann** [comment]
So far what I observe is the gradle templates that were generated from the CI:
Screenshot from 2022-08-10 11-53-36
Screenshot from 2022-08-10 11-53-21

**Comment by atomfrede** [comment]
I guess the Gradle plugin configures native Image build now. We should not mix it up with the native build issue. The ci is not cpbfigured to build native images.  Maybe we need to adapt the build commands?

**Comment by atomfrede** [comment]
I guess the Gradle plugin configures native Image build now. We should not mix it up with the native build issue. The ci is not cpbfigured to build native images.  Maybe we need to adapt the build commands?

File: generators/server/templates/src/main/resources/application-prod.yml.ejs
Comment (atomfrede) [inline generators/server/templates/src/main/resources/application-prod.yml.ejs:132]: Mostly not relevant for the ci pipeline, but we should keep `<%= serverPort %>` it is supposed to come from the main generator

File: generators/server/templates/src/main/resources/application-dev.yml.ejs
Comment (atomfrede) [inline generators/server/templates/src/main/resources/application-dev.yml.ejs:106]: Mostly not relevant for the ci pipeline, but we should keep <%= serverPort %> it is supposed to come from the main generator

File: generators/constants.cjs
Comment (atomfrede) [inline generators/constants.cjs:66]:
```suggestion
    testcontainers: '1.17.3',
```

File: generators/constants.cjs
Comment (atomfrede) [inline generators/constants.cjs:64]:
```suggestion
    archunit: '0.23.1',
```

File: generators/constants.cjs
Comment (atomfrede) [inline generators/constants.cjs:63]:
```suggestion
    caffeine: '3.1.1',
```

**Comment by atomfrede** [comment]
Thanks @kevintanhongann for keep up on this. I will have some time next week to help test and getting this done. Great work so far 👍

**Comment by kevintanhongann** [comment]
@atomfrede I have to pass the baton to one of you guys to fix the maven portion of the issue, aside from the CorsTest testing (in which I'm not sure what to look out for).

Next thing is to merge #252 in which I assume to have merged in main already, in which it completes the user registration workflow cycle.

Thanks again.

**Comment by mraible** [comment]
@atomfrede Are you able to take a look at this? I can take a look next week if you don't have time.

**Comment by atomfrede** [comment]
@mraible I won't have time until next wednesday, maybe we can join forces or so. My approach would be to pick one configuration (e.g. mvn + jwt + angular) and try to make it work.

File: generators/server/templates/src/main/java/package/security/DatabaseAuthenticationProvider.java.ejs
Comment (atomfrede) [inline generators/server/templates/src/main/java/package/security/DatabaseAuthenticationProvider.java.ejs:12]:
```suggestion
import <%=packageName%>.repository.AuthorityRepository;
```

**Comment by kevintanhongann** [comment]
After all the changes we have done, and was testing with a prototype I was working on, it seems to me that the user login wasn't cached properly for some reason. 

SecurityUtil.getCurrentUserLogin - no user present 

userService.getUserWithAuthorities by login name or email - no user present as well. 

Still investigating what's wrong with the cache configuration.

**Comment by mraible** [comment]
I tried to use this branch to create a basic application with `mhipster jdl default`. It generates OK, but has the following error when I start it:

```
 __  __ _                                  _
|  \/  (_) ___ _ __ ___  _ __   __ _ _   _| |_
| |\/| | |/ __| '__/ _ \| '_ \ / _` | | | | __|
| |  | | | (__| | | (_) | | | | (_| | |_| | |_
|_|  |_|_|\___|_|  \___/|_| |_|\__,_|\__,_|\__|
  Micronaut (v3.6.0)

00:48:58.107 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Could not initialize class io.micronaut.inject.annotation.AnnotationMetadataSupport
java.lang.NoClassDefFoundError: Could not initialize class io.micronaut.inject.annotation.AnnotationMetadataSupport
	at io.micronaut.inject.annotation.DefaultAnnotationMetadata.findRepeatableAnnotation(DefaultAnnotationMetadata.java:1427)
	at io.micronaut.inject.annotation.DefaultAnnotationMetadata.findRepeatableAnnotation(DefaultAnnotationMetadata.java:1419)
	at io.micronaut.core.annotation.AnnotationMetadata.hasStereotype(AnnotationMetadata.java:1455)
	at io.micronaut.context.AbstractBeanContextConditional.isEnabled(AbstractBeanContextConditional.java:46)
	at io.micronaut.context.AbstractBeanConfiguration.isEnabled(AbstractBeanConfiguration.java:30)
	at io.micronaut.inject.BeanContextConditional.isEnabled(BeanContextConditional.java:40)
	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:3281)
	at io.micronaut.context.DefaultBeanContext.finalizeConfiguration(DefaultBeanContext.java:3684)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:341)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:190)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:75)
	at com.mycompany.myapp.JhipsterApp.main(JhipsterApp.java:69)
```

**Comment by mraible** [comment]
@JasonTypesCodes Do you have any idea why [this error](https://github.com/jhipster/generator-jhipster-micronaut/pull/267#issuecomment-1286533410) is happening?

**Comment by mraible** [comment]
Is there anything I can do to help move this along?

**Comment by mraible** [comment]
@kevintanhongann @atomfrede Do y'all have time to work on this pull request? If not, I'd be happy to help with some guidance.

**Comment by atomfrede** [comment]
Sorry @mraible at the moment I always fail to code something in the evenings. But I am still interested to make this work. Hopefully when the days are getting longer I have more energy.

**Comment by kevintanhongann** [comment]
I'll get back to this asap. Sorry for the long wait. I had a project that I
was working on to troubleshoot the case but didn't get back to it because I
had a busy schedule too.

Perhaps we will need to schedule a call so that I can share my findings for
us to proceed with this. What do you guys think?

**Comment by mraible** [comment]
> Perhaps we will need to schedule a call so that I can share my findings for us to proceed with this.

In my 20+ years of contributing to open source, I've never had a call to discuss any issues or pull requests. I'm not about to start now. 😅

I think it's best to share your findings here, so everyone can see them and offer advice for how to fix them.

**Comment by madhutst** [comment]
> I tried to use this branch to create a basic application with `mhipster jdl default`. It generates OK, but has the following error when I start it:
> 
> ```
>  __  __ _                                  _
> |  \/  (_) ___ _ __ ___  _ __   __ _ _   _| |_
> | |\/| | |/ __| '__/ _ \| '_ \ / _` | | | | __|
> | |  | | | (__| | | (_) | | | | (_| | |_| | |_
> |_|  |_|_|\___|_|  \___/|_| |_|\__,_|\__,_|\__|
>   Micronaut (v3.6.0)
> 
> 00:48:58.107 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Could not initialize class io.micronaut.inject.annotation.AnnotationMetadataSupport
> java.lang.NoClassDefFoundError: Could not initialize class io.micronaut.inject.annotation.AnnotationMetadataSupport
> 	at io.micronaut.inject.annotation.DefaultAnnotationMetadata.findRepeatableAnnotation(DefaultAnnotationMetadata.java:1427)
> 	at io.micronaut.inject.annotation.DefaultAnnotationMetadata.findRepeatableAnnotation(DefaultAnnotationMetadata.java:1419)
> 	at io.micronaut.core.annotation.AnnotationMetadata.hasStereotype(AnnotationMetadata.java:1455)
> 	at io.micronaut.context.AbstractBeanContextConditional.isEnabled(AbstractBeanContextConditional.java:46)
> 	at io.micronaut.context.AbstractBeanConfiguration.isEnabled(AbstractBeanConfiguration.java:30)
> 	at io.micronaut.inject.BeanContextConditional.isEnabled(BeanContextConditional.java:40)
> 	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:3281)
> 	at io.micronaut.context.DefaultBeanContext.finalizeConfiguration(DefaultBeanContext.java:3684)
> 	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:341)
> 	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:190)
> 	at io.micronaut.runtime.Micronaut.start(Micronaut.java:75)
> 	at com.mycompany.myapp.JhipsterApp.main(JhipsterApp.java:69)
> ```

Do you find solution to this, I had the same issue after changing several dependency versions.

**Comment by wund3rcr4zy** [comment]
> > I tried to use this branch to create a basic application with `mhipster jdl default`. It generates OK, but has the following error when I start it:
> > ```
> >  __  __ _                                  _
> > |  \/  (_) ___ _ __ ___  _ __   __ _ _   _| |_
> > | |\/| | |/ __| '__/ _ \| '_ \ / _` | | | | __|
> > | |  | | | (__| | | (_) | | | | (_| | |_| | |_
> > |_|  |_|_|\___|_|  \___/|_| |_|\__,_|\__,_|\__|
> >   Micronaut (v3.6.0)
> > 
> > 00:48:58.107 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Could not initialize class io.micronaut.inject.annotation.AnnotationMetadataSupport
> > java.lang.NoClassDefFoundError: Could not initialize class io.micronaut.inject.annotation.AnnotationMetadataSupport
> > 	at io.micronaut.inject.annotation.DefaultAnnotationMetadata.findRepeatableAnnotation(DefaultAnnotationMetadata.java:1427)
> > 	at io.micronaut.inject.annotation.DefaultAnnotationMetadata.findRepeatableAnnotation(DefaultAnnotationMetadata.java:1419)
> > 	at io.micronaut.core.annotation.AnnotationMetadata.hasStereotype(AnnotationMetadata.java:1455)
> > 	at io.micronaut.context.AbstractBeanContextConditional.isEnabled(AbstractBeanContextConditional.java:46)
> > 	at io.micronaut.context.AbstractBeanConfiguration.isEnabled(AbstractBeanConfiguration.java:30)
> > 	at io.micronaut.inject.BeanContextConditional.isEnabled(BeanContextConditional.java:40)
> > 	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:3281)
> > 	at io.micronaut.context.DefaultBeanContext.finalizeConfiguration(DefaultBeanContext.java:3684)
> > 	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:341)
> > 	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:190)
> > 	at io.micronaut.runtime.Micronaut.start(Micronaut.java:75)
> > 	at com.mycompany.myapp.JhipsterApp.main(JhipsterApp.java:69)
> > ```
> 
> Do you find solution to this, I had the same issue after changing several dependency versions.

I've run into this exact error recently and found this post while searching for solutions. For me it turned out to be that jakarta annotations were not available.

```
java.lang.NoClassDefFoundError: jakarta/annotation/PreDestroy
	at io.micronaut.inject@3.8.4/io.micronaut.inject.annotation.AnnotationMetadataSupport.<clinit>(AnnotationMetadataSupport.java:63)
	at gordianknot.cli/com.wund3rcr4zy.gordianknot.cli.Application.main(Application.java:12)
Caused by: java.lang.ClassNotFoundException: jakarta.annotation.PreDestroy
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 2 more
```

Switching my module-info.java from 'requires static jakarta.annotation;' to 'requires jakarta.annotation' solved the above error.

I hope this helps in your case.

**Comment by mraible** [comment]
I don't see any `module-info.java` files in this blueprint, so I'm not sure that will fix the problem. I tried `mhipster jdl default` today with Node 18 and it fails with the following error:

```
Unhandled promise rejection at:
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './generators/generator-constants' is not defined
by "exports" in /Users/matt.raible/dev/generator-jhipster-micronaut/node_modules/generator-jhipster/package.json
```

Using Node 16 works. I also noticed there seems to be duplicate javax and jakarta dependencies. 

Screenshot 2023-02-14 at 9 37 08 AM

If I run `mvn dependency:tree`, these seem to be caused by Micronaut dependencies, so I'm not sure if jakarta or javax dependencies are correct.

File: generators/constants.cjs
Comment (mraible) [inline generators/constants.cjs:47]:
```suggestion
    micronautData: '3.10.0',
```

**Comment by mraible** [comment]
@kevintanhongann I tried upgrading this PR to use Micronaut 3.9.1, but I'm unable to commit my changes.

**Comment by kevintanhongann** [comment]
@mraible in which option do I allow for your commits?

**Comment by mraible** [comment]
I accepted the invite, but it still doesn't let me commit my suggestions. Maybe you can do it?

**Comment by kevintanhongann** [comment]
@mraible do try again. I forgot that I enabled branch protection for my own fork.

**Comment by mraible** [comment]
@kevintanhongann It looks like it worked! We'll see what CI says. I'm not certain these changes will fix anything.

**Comment by kevintanhongann** [comment]
@mraible i'll pull your changes and cross reference with my sample project to see what needs to be improved on.

**Comment by kevintanhongann** [comment]
@mraible I've added you to a sample prototype that I'm working on. I'm still investigating why the SecurityUtils.getCurrentUserLogin is still not returning anything even though I'm logged in from the UI. If there's a way to debug the cache in the Spring Boot cache configuration, do lemme know.

Login with **user@localhost** and password **user** and look at the dashboard endpoint at line 62.

**Comment by kevintanhongann** [comment]
From the sample prototype, I was debugging whether or not the cache name exists.
Apparently cache names **usersByEmail** and **usersByEmailOrLogin** actually exists, but SecurityUtils.getCurrentUserLogin didn't manage to get it.

```
public static Optional<String> getCurrentUserLogin() {
    return ServerRequestContext
      .currentRequest()
      .flatMap(request -> request.getUserPrincipal(Authentication.class))
      .map(Authentication::getName);
  }
```

What do you think we should do to correct this bit?

**Comment by jeremyg484** [comment]
Hi, all! I am a member of the Micronaut dev team and we're looking at getting things up-to-date for JHipster now that Micronaut Framework 4 has been released.

I think getting this PR over the finish line will be a great first step, and I'll do whatever I can to help. I'm starting to examine it now to figure out what is broken. If there's anything specific anyone wants me to look at aside from the most recent points discussed above, please do let me know.

**Comment by jeremyg484** [comment]
I created my own fork based on the great work already done here by @kevintanhongann and @mraible and I believe I've at least got all of the basics working now for generating both Maven and Gradle apps based on Micronaut framework 3.

Since I'm not a maintainer here, it seemed the easiest approach would be to submit a PR to @kevintanhongann's fork, which I have done here:
https://github.com/kevintanhongann/generator-jhipster-micronaut/pull/1

If anyone has an alternate / preferred way to go about it, please let me know.

I still want to do some more testing to make sure I haven't missed anything, and then I've got some miscellaneous (but not critical) improvements in mind, but I'd like to just start here with having a working version merged back into main that can be used as a basis for those further enhancements. Anyone else who wants to give it a try and let me know if they spot any issues, that would be great as well.

My thought is that I can go ahead and use this to get a Micronaut framework 4 version going based on the same code. From what I've spotted so far, that should only require some minor tweaks to the Maven and Gradle build configs.

**Comment by kevintanhongann** [comment]
@jeremyg484 might need you to sign the CLA on your end first before we do anything else. Good work so far.

**Comment by kevintanhongann** [comment]
@jeremyg484 I suppose I agree that we can bump the versioning to Micronaut 4 and see what happens. Technically there should not be any major issues. Thoughts everyone?

**Comment by jeremyg484** [comment]
> @jeremyg484 I suppose I agree that we can bump the versioning to Micronaut 4 and see what happens. Technically there should not be any major issues. Thoughts everyone?

My thought is actually to keep that in a separate branch for now. Micronaut framework 4 requires Java 17 as a baseline, so there is probably value in still providing support for generating applications that use v3.

I wrote most of the Micronaut 3-to-4 OpenRewrite upgrade recipe, so know from experience it's not as simple as just bumping the version. There are other adjustments that must be made to mainly build.gradle / pom.xml due to a slightly different structure to all of the platform dependencies.

If we were to want to support both v3 and v4, how would that typically be handled in JHipster? I'm guessing two differently versioned releases of the blueprint?

**Comment by kevintanhongann** [comment]
@jeremyg484 as far as I observed for jHipster Spring Boot and the Micronaut variant, versioning seems pretty static and jHipster version upgrades along with the backend versions, and you cant specify the spring boot or micronaut version(in this case) that you want.

**Comment by jeremyg484** [comment]
@kevintanhongann The blueprint is ultimately just distributed as an NPM module, right? So it would appear to me in theory, there could be say a v2.0.0 of the module that generates Micronaut framework 3 apps, and a v3.0.0 that generates Micronaut framework 4 apps and then it's just a matter of giving users the right `npm install` instructions for their preferred path.

I realize that certainly complicates things as far as maintenance of this repo. I'll leave it up to you and the project maintainers to voice whether the extra complication to support both would be worthwhile for the JHipster community.

My hope is that I can have an alternate Micronaut framework 4 version of this current code pushed up to my own repo my the end of the day, then you all could take a look at the differences to decide the preferred approach.

**Comment by jeremyg484** [comment]
A quick update - since the CI workflows haven't executed yet (looks like they're waiting on approval?), I dug in and did some extensive running and testing of the samples myself. Turns out a couple of them are still broken, which isn't surprising as they include options I hadn't manually tested yet. In particular, the Redis sample and the OAuth samples seem to be breaking. I'm still debugging, but will hopefully be able to push more changes to fix those tomorrow.

**Comment by atomfrede** [comment]
Sorry @jeremyg484 for the late reply. You first wrote on my first day of vacation and I totally forget about it. And I think @mraible is on vacation too. So big thanks for taking care of the micronaut update that's awesome. I would say we just need to support micronaut 4. The next jhipster release 8, which is in beta right now will require Java 17 so no need to create and to support multiple versions of the blueprint.

**Comment by jeremyg484** [comment]
@atomfrede No worries! That sounds reasonable to me. I'll go ahead and start shifting things to use Micronaut framework 4 alongside working out these remaining issues.

**Comment by jeremyg484** [comment]
I have a working branch that takes the existing state of this PR and updates everything to Micronaut 4.

https://github.com/jeremyg484/generator-jhipster-micronaut/commits/micronaut-4

As before, the basics are working with manual testing, but I still need to dig in and ensure it passes the CI tests (at the moment it does not).

Once I have that working, how would everyone prefer I proceed? Should I continue pushing to this same PR branch (and we change the title of the PR to "Upgrade micronaut 2 to 4")? That seems like it would be the most straightforward, but I just wanted to check.

**Comment by jeremyg484** [comment]
@kevintanhongann, @atomfrede, @mraible

Ok, so after some deeper investigation I realize my micronaut-4 branch is missing something fairly crucial, because to this point I'd only been testing it with `mhipster jdl default` and `mhipster jdl default-gradle`. I've not done anything yet to modify the way entities are generated such that they use jakarta.* packages instead of javax.* packages. Micronaut framework 4 has transitioned completely to using only jakarta.* packages. So while the base functionality works, the next step beyond that (generating actual domain classes) fails at the moment because everything there is still getting generated with javax.* package imports.

I believe I can see where I could hook in and change the package imports, but is it worth it? Should the Micronaut framework 4 version of this instead be done in a completely separate PR targeting JHipster 8, where that transition already seems to be done?

Your opinions as to the best direction here would be most welcome.

(FYI, I am about to be on vacation for a week myself, so probably won't make much more progress until I return next Tuesday.)

**Comment by mraible** [comment]
@jeremyg484 Sorry for the delay; I was having a blast on summer vacation. 

Ideally, we could merge this PR once everything succeeds and release v2.0 with Micronaut 3 support. Then, merge the next PR to add Micronaut 4 support and release 3.0. 

That way, @kevintanhongann could get credit for the work he's done for Micronaut 3.  The PR for Micronaut 4 might be helpful for showing folks how to migrate from 3 to 4.

**Comment by jeremyg484** [comment]
> @jeremyg484 Sorry for the delay; I was having a blast on summer vacation.

No worries, I'm about to do the same. Looks like we were typing simultaneously. :) (See my comment just above re: Micronaut 4.)

> Ideally, we could merge this PR once everything succeeds and release v2.0 with Micronaut 3 support. Then, merge the next PR to add Micronaut 4 support and release 3.0.
> 
> That way, @kevintanhongann could get credit for the work he's done for Micronaut 3. 
I agree, that makes sense. I think to get this particular PR done, it will just require getting the Redis and OAuth tests passing, which is probably a lighter lift at this point overall.

> The PR for Micronaut 4 might be helpful for showing folks how to migrate from 3 to 4.
Makes sense. I've pushed up the currently generated default samples - they show what is required beyond just running the OpenRewrite recipe (it did do most of the work, but there are some additional dependency conflicts and such that I had to work out in this case).

They are here if anyone wants to take a look - the most notable differences are in the build files:

https://github.com/jeremyg484/jdl-default-maven-micronaut4

https://github.com/jeremyg484/jdl-default-gradle-micronaut4

I kinda sounds like my instinct is right and that we should align Micronaut  support with JHipster 8.

**Comment by mraible** [comment]
> It kinda sounds like my instinct is right and that we should align Micronaut support with JHipster 8.

I agree. I'll work on getting this PR to pass all tests this week.

**Comment by mraible** [comment]
Tests are failing with:

> Couldn't find type javax.inject.Singleton. Are you missing a dependency on your classpath?

I'll try adding:

```xml
<dependency>
    <groupId>javax.inject</groupId>
    <artifactId>javax.inject</artifactId>
    <version>1</version>
</dependency>
```

File: generators/server/templates/pom.xml.ejs
Comment (jeremyg484) [inline generators/server/templates/pom.xml.ejs:233]:
```suggestion
            <version>${javax-inject.version}</version>
```

Comment (jeremyg484) [reply generators/server/templates/pom.xml.ejs:233]: Gets things further along to where we can at least see some specific failing tests. :)

Out of curiosity, know what is requiring the javax.inject dependency? It's not any of the generated code and it's not anything from Micronaut.

Comment (mraible) [reply generators/server/templates/pom.xml.ejs]: It was a failure in CI: https://github.com/jhipster/generator-jhipster-micronaut/actions/runs/5859326680/job/15885088988

**Comment by jeremyg484** [comment]
I just sent a PR for a couple other minor fixes to @kevintanhongann's branch. It doesn't resolve the Redis or OAuth tests, but fixes an error I noticed in manually testing the `jdl-default-app` CI sample and also gets import of Maven projects into IntelliJ working correctly by updating the version of MapStruct.

https://github.com/kevintanhongann/generator-jhipster-micronaut/pull/2

**Comment by mraible** [comment]
Wahoo! All tests pass, so I'm merging this.

**Comment by kevintanhongann** [comment]
Awesome job everyone. Glad to see things moving at a faster pace.

**Pull request merged**
